### PR TITLE
refactor: change metadata system contract and add annotation

### LINKED ISSUE #1441: System contracts `ckb_light_client` and `image_cell` are saved into different MPT trie.
### Current Behavior

- Initialize system contracts when Axon starts.

  https://github.com/axonweb3/axon/blob/8e5d5672080cc3b3bb784af08dc1c54ed5adbe43/core/run/src/lib.rs#L480

- Generate MPT changes.

  https://github.com/axonweb3/axon/blob/8e5d5672080cc3b3bb784af08dc1c54ed5adbe43/core/executor/src/system_contract/mod.rs#L133-L136

- Two changes are created.

  https://github.com/axonweb3/axon/blob/8e5d5672080cc3b3bb784af08dc1c54ed5adbe43/core/executor/src/system_contract/utils.rs#L73-L88

- Two changes are applied one by one.

  https://github.com/axonweb3/axon/blob/8e5d5672080cc3b3bb784af08dc1c54ed5adbe43/core/executor/src/adapter/backend/apply.rs#L235-L248

- Since there two changes have different addresses.

  - Each time a new `old_account` is initialized.

    https://github.com/axonweb3/axon/blob/8e5d5672080cc3b3bb784af08dc1c54ed5adbe43/core/executor/src/adapter/backend/apply.rs#L155-L163

  - Then, each time a new `storage_trie` is initialized.

    https://github.com/axonweb3/axon/blob/8e5d5672080cc3b3bb784af08dc1c54ed5adbe43/core/executor/src/adapter/backend/apply.rs#L165-L169
    https://github.com/axonweb3/axon/blob/8e5d5672080cc3b3bb784af08dc1c54ed5adbe43/core/executor/src/adapter/backend/apply.rs#L171-L175

- So, these two system contracts are insert into two different MPT trie.

- Let's go back to the system contracts initialization.

  - When try to get the `current_cell_root`, only address `CKB_LIGHT_CLIENT_CONTRACT_ADDRESS` is used.

    https://github.com/axonweb3/axon/blob/8e5d5672080cc3b3bb784af08dc1c54ed5adbe43/core/executor/src/system_contract/mod.rs#L130-L131

    :question: So, its `storage_root` won't contains the image cell contract. I'm not sure. Is this correct?

  - When try to generate MPT changes, only address `IMAGE_CELL_CONTRACT_ADDRESS` is used.

    https://github.com/axonweb3/axon/blob/8e5d5672080cc3b3bb784af08dc1c54ed5adbe43/core/executor/src/system_contract/mod.rs#L134
   
    :question: Can anyone guess the fact this function call also generate changes for the ckb light client contract too, even just an address `IMAGE_CELL_CONTRACT_ADDRESS` is passed.

  - :question: Can anyone guess that the above two function calls are a pair: one is getter, and one is setter, even they use different addresses as arguments.

### Expected Behavior

.

### Axon version

main branch

### LINKED ISSUE #1457: Call the built-in `Metadata` contract causes an error.
### Summary

Call the built-in `Metadata` contract causes an error:

> CKB-VM error execution reverted

### How to reproduce?

- Create a interface file for `Metadata` contract.

  - Base on [the built-in `Metadata` contract SOL file].

  - Base on [the internal `Metadata` contract ABI file].

- Call the method `getMetadata(uint64 epoch)` from another contract.

  ```solidity
  function getMetadata(uint64 epoch) public view returns (IMetadataManager.Metadata memory){ 
    address addr = 0xFFfffFFfFFfffFfFffFFfFfFfFffFfffFFFFFf01;
    return IMetadataManager(addr).getMetadata(epoch);
  }
  ```

  References:

  - Method signature in [the built-in `Metadata` contract SOL file].

    https://github.com/axonweb3/axon/blob/71f5bad3ae581108b708e671c3e3573af4564dde/builtin-contract/metadata/contracts/metadata.sol#L139

  - Method signature in [the internal `Metadata` contract ABI file].

    https://github.com/axonweb3/axon/blob/71f5bad3ae581108b708e671c3e3573af4564dde/core/executor/src/system_contract/metadata/abi/metadata_abi.json#L141-L148

There are several issues:

- [x] Why the error message contains "CKB-VM"?

  I didn't call any contract which uses CKB-VM.

- [x] How to call `Metadata` contract in Solidity scripts with the provided method?

  The provided method is:

  https://github.com/axonweb3/axon/blob/71f5bad3ae581108b708e671c3e3573af4564dde/core/executor/src/precompiles/metadata.rs#L11-L17

  p.s. Axon should NOT force users to write assembly.

- [x] Should there be some tests before new changes on the built-in `Metadata` contract? (ref: #1454)

  Did I miss some things?
  I didn't find any tests under [the `test` directory](https://github.com/axonweb3/axon/tree/71f5bad3ae581108b708e671c3e3573af4564dde/tests) for the built-in `Metadata` contract.

  I'm neither an expert on JavaScript nor an expert on Solidity.
  Also, I haven't understood the storage model of Axon completely, yet.

  I hope someone can add such tests (or examples) to increase confidence for the changes of the built-in `Metadata` contract.

[the built-in `Metadata` contract SOL file]: https://github.com/axonweb3/axon/blob/71f5bad3ae581108b708e671c3e3573af4564dde/builtin-contract/metadata/contracts/metadata.sol
[the internal `Metadata` contract ABI file]: https://github.com/axonweb3/axon/blob/71f5bad3ae581108b708e671c3e3573af4564dde/core/executor/src/system_contract/metadata/abi/metadata_abi.json

- [x] Provide an interface SOL file for the built-in `Metadata` contract, and use unit tests to make sure it will be updated when the contract is changed.

  Axon should NOT let users to write the interface manually.

  It's better than provide a contract SOL file without code for implementation. (ref: #1449)

### Expected Behavior

.

### PR DESCRIPTION
**Pull request by KaoImin**
## What this PR does / why we need it?

1. Change the built-in solidity contracts into interface definition. Because they only define the interface of system contracts which is implemented in Rust.
2. Use the same address in system contact init avoiding the follow situation:
https://github.com/axonweb3/axon/blob/8e5d5672080cc3b3bb784af08dc1c54ed5adbe43/core/executor/src/system_contract/mod.rs#L130-L131
https://github.com/axonweb3/axon/blob/8e5d5672080cc3b3bb784af08dc1c54ed5adbe43/core/executor/src/system_contract/mod.rs#L134
3. Remove the `metadata` precompile contract

### What is the impact of this PR?

No Breaking Change

**PR relation**:
- Ref #1441 

<details><summary>CI Settings</summary><br/>

### **CI Usage**

**Tip**: Check the CI you want to run below, and then comment `/run-ci`.

**CI Switch**

- [x] Web3 Compatible Tests
- [x] OCT 1-5 And 12-15
- [x] OCT 6-10
- [x] OCT 11
- [x] OCT 16-19
- [x] v3 Core Tests

### **CI Description**

| CI Name                                   | Description                                                               |
| ----------------------------------------- | ------------------------------------------------------------------------- |
| *Web3 Compatible Test*                    | Test the Web3 compatibility of Axon                                       |
| *v3 Core Test*                            | Run the compatibility tests provided by Uniswap V3                        |
| *OCT 1-5 \| 6-10 \| 11 \| 12-15 \| 16-19* | Run the compatibility tests provided by OpenZeppelin                      |

</details>

### CODE REVIEW

File: builtin-contract/metadata/contracts/metadata.sol [hunk 1]
```diff
@@ -2,7 +2,9 @@
 
 pragma solidity >=0.7.0;
 
-// import "hardhat/console.sol";
+// **Notice**
+// The solidity contract only define the interface of metadata contract. The real
+// implementation is in `core/executor/src/system_contract/metadata`.
```
Comment (Flouse) [inline builtin-contract/metadata/contracts/metadata.sol]:
```suggestion
// **Notice**
// This solidity contract only defines the interface of metadata contract. The real
// implementation is in `core/executor/src/system_contract/metadata`.
```

**Review by chaoticlonghair — COMMENTED** [review]
Why don't just write an interface?

I don't think a broken[^1] contract is better than an interface.

Users can use the interface. (ref: #1457)

[^1]: Not able to compile.

File: core/executor/src/tests/system_script/metadata.rs [hunk 2]
```diff
@@ -185,7 +187,6 @@ fn prepare_metadata() -> Metadata {
         propose_counter:  vec![],
         consensus_config: ConsensusConfig {
             gas_limit:       1u64,
-            gas_price:       0u64,
```
Comment (Flouse) [inline core/executor/src/tests/system_script/metadata.rs:188]: Please delete the `gas_price` consensus_config in `**/chain-spec.toml` as well.

**Review by Flouse — DISMISSED** [review]

**Review by Flouse — APPROVED** [review]

File: core/executor/src/tests/system_script/metadata.rs [hunk 3]
```diff
@@ -47,6 +48,7 @@ fn test_init<'a>(backend: &mut MemoryBackend<'a>, executor: &MetadataContract<Me
     let addr = H160::from_str("0xf000000000000000000000000000000000000000").unwrap();
     let tx = prepare_tx_1(&addr);
     let r = executor.exec_(backend, &tx);
+    println!("{:?}", r);
```
Comment (chaoticlonghair) [inline core/executor/src/tests/system_script/metadata.rs:51]: I think the previous `env_logger::init();` and this line is not matched.

Each format is fine, but mix two formats is strange。

**Review by chaoticlonghair — APPROVED** [review]
I have different opinions, but since the deadline is already reached, I don't want to block the project progress.

File: core/executor/src/tests/system_script/metadata.rs (hunk 3 above)
Comment (Flouse) [reply core/executor/src/tests/system_script/metadata.rs:51]: - recorded in https://github.com/axonweb3/axon/issues/1464

**Pull request merged**